File: src/test/java/no/nav/syfo/mappers/SykmeldingMapperTest.java
package no.nav.syfo.mappers;

import no.nav.syfo.controller.domain.sykmelding.Periode;
import no.nav.syfo.controller.domain.sykmelding.Sykmelding;
import no.nav.syfo.testutils.SykmeldingMocks;
import no.nav.tjeneste.virksomhet.sykmelding.v1.informasjon.*;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import java.time.LocalDateTime;
import java.util.List;

import static java.time.LocalDate.of;
import static no.nav.syfo.mappers.SykmeldingMapper.sykmelding;
import static org.assertj.core.api.Assertions.assertThat;

@RunWith(MockitoJUnitRunner.class)
public class SykmeldingMapperTest {

    @Test
    public void testMappingenFraWebserviceobjektene() throws Exception {
        Sykmelding sykmelding = sykmelding(meldingWS());

        assertThat(sykmelding.id).isNotNull();
        assertThat(sykmelding.arbeidsgiver).isNotEmpty();
        assertThat(sykmelding.diagnose.hoveddiagnose).isNotNull();
        assertThat(sykmelding.pasient.etternavn).isNotEmpty();
        assertThat(sykmelding.pasient.fnr).isNotEmpty();
        assertThat(sykmelding.pasient.fornavn).isNotEmpty();
        assertThat(sykmelding.bekreftelse.sykmelder).isNotEmpty();

        assertThat(sykmelding.diagnose.fravaersgrunnLovfestet).isNotEmpty();
        assertThat(sykmelding.diagnose.fravaerBeskrivelse).isNotEmpty();

        assertThat(sykmelding.mulighetForArbeid.perioder).isNotEmpty();
        List<Periode> perioder = sykmelding.mulighetForArbeid.perioder;
        Periode aktivitetIkkeMulig = perioder.get(0);
        Periode gradert = perioder.get(1);
        Periode behandlingsdager = perioder.get(2);
        Periode reisetilskudd = perioder.get(3);
        Periode avventende = perioder.get(4);

        assertThat(aktivitetIkkeMulig.fom).isNotNull();
        assertThat(aktivitetIkkeMulig.tom).isNotNull();
        assertThat(aktivitetIkkeMulig.grad).isEqualTo(100);
        assertThat(gradert.grad).isGreaterThan(0).isLessThan(100);
        assertThat(behandlingsdager.behandlingsdager).isGreaterThan(0);
        assertThat(reisetilskudd.reisetilskudd).isTrue();
        assertThat(avventende.avventende).isNotEmpty();

        assertThat(sykmelding.mulighetForArbeid.aktivitetIkkeMulig433).isNotEmpty();
        assertThat(sykmelding.mulighetForArbeid.aarsakAktivitetIkkeMulig433).isNotEmpty();

        assertThat(sykmelding.bekreftelse.utstedelsesdato).isNotNull();

        assertThat(sykmelding.sporsmal.arbeidssituasjon).isEqualTo("ARBEIDSTAKER");
        assertThat(sykmelding.sporsmal.fravaersperioder.get(0).fom).isEqualTo(of(2018, 10, 17));
        assertThat(sykmelding.sporsmal.fravaersperioder.get(0).tom).isEqualTo(of(2018, 10, 17));

    }

    private WSMelding meldingWS() throws Exception {
        return new WSMelding()
                .withMeldingId("1")
                .withStatus("NY")
                .withSendtTilArbeidsgiverDato(LocalDateTime.of(2016, 4, 4, 12, 40))
                .withIdentdato(of(2016, 4, 4))
                .withSykmelding(SykmeldingMocks.getWSSykmelding())
                .withSmSpoersmaal(
                        new WSSMSpoersmaal()
                                .withHarAnnetFravaer(true)
                                .withForsikringsgrad(60)
                                .withHarForsikringsgrad(true)
                                .withArbeidssituasjon(WSArbeidssituasjon.ARBEIDSTAKER)
                                .withAnnenFravaersperiodeListe(new WSDatospenn()
                                        .withFom(of(2018, 10, 17))
                                        .withTom(of(2018, 10, 17))));
    }


}

File: src/test/kotlin/no/nav/syfo/mappers/SykmeldingMapperTest.kt
package no.nav.syfo.mappers

import no.nav.syfo.testhelper.SykmeldingMocks.getWSSykmelding
import no.nav.tjeneste.virksomhet.sykmelding.v1.informasjon.*
import org.assertj.core.api.Assertions
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.junit.MockitoJUnitRunner
import java.time.LocalDate
import java.time.LocalDateTime

@RunWith(MockitoJUnitRunner::class)
class SykmeldingMapperTest {
    @Test
    @Throws(Exception::class)
    fun testMappingenFraWebserviceobjektene() {
        val sykmelding = SykmeldingMapper.sykmelding(meldingWS())

        Assertions.assertThat(sykmelding.id).isNotNull()
        Assertions.assertThat(sykmelding.arbeidsgiver).isNotEmpty()
        Assertions.assertThat(sykmelding.diagnose.hoveddiagnose).isNotNull
        Assertions.assertThat(sykmelding.pasient.etternavn).isNotEmpty()
        Assertions.assertThat(sykmelding.pasient.fnr).isNotEmpty()
        Assertions.assertThat(sykmelding.pasient.fornavn).isNotEmpty()
        Assertions.assertThat(sykmelding.bekreftelse.sykmelder).isNotEmpty()
        Assertions.assertThat(sykmelding.diagnose.fravaersgrunnLovfestet).isNotEmpty()
        Assertions.assertThat(sykmelding.diagnose.fravaerBeskrivelse).isNotEmpty()
        Assertions.assertThat(sykmelding.mulighetForArbeid.perioder).isNotEmpty

        val perioder = sykmelding.mulighetForArbeid.perioder
        val aktivitetIkkeMulig = perioder[0]
        val gradert = perioder[1]
        val behandlingsdager = perioder[2]
        val reisetilskudd = perioder[3]
        val avventende = perioder[4]

        Assertions.assertThat(aktivitetIkkeMulig.fom).isNotNull()
        Assertions.assertThat(aktivitetIkkeMulig.tom).isNotNull()
        Assertions.assertThat(aktivitetIkkeMulig.grad).isEqualTo(100)
        Assertions.assertThat(gradert.grad).isGreaterThan(0).isLessThan(100)
        Assertions.assertThat(behandlingsdager.behandlingsdager).isGreaterThan(0)
        Assertions.assertThat(reisetilskudd.reisetilskudd).isTrue()
        Assertions.assertThat(avventende.avventende).isNotEmpty()
        Assertions.assertThat(sykmelding.mulighetForArbeid.aktivitetIkkeMulig433).isNotEmpty
        Assertions.assertThat(sykmelding.mulighetForArbeid.aarsakAktivitetIkkeMulig433).isNotEmpty()
        Assertions.assertThat(sykmelding.bekreftelse.utstedelsesdato).isNotNull()
        Assertions.assertThat(sykmelding.sporsmal.arbeidssituasjon).isEqualTo("ARBEIDSTAKER")
        Assertions.assertThat(sykmelding.sporsmal.fravaersperioder[0].fom).isEqualTo(LocalDate.of(2018, 10, 17))
        Assertions.assertThat(sykmelding.sporsmal.fravaersperioder[0].tom).isEqualTo(LocalDate.of(2018, 10, 17))
    }

    @Throws(Exception::class)
    private fun meldingWS(): WSMelding {
        return WSMelding()
            .withMeldingId("1")
            .withStatus("NY")
            .withSendtTilArbeidsgiverDato(LocalDateTime.of(2016, 4, 4, 12, 40))
            .withIdentdato(LocalDate.of(2016, 4, 4))
            .withSykmelding(getWSSykmelding())
            .withSmSpoersmaal(
                WSSMSpoersmaal()
                    .withHarAnnetFravaer(true)
                    .withForsikringsgrad(60)
                    .withHarForsikringsgrad(true)
                    .withArbeidssituasjon(WSArbeidssituasjon.ARBEIDSTAKER)
                    .withAnnenFravaersperiodeListe(WSDatospenn()
                        .withFom(LocalDate.of(2018, 10, 17))
                        .withTom(LocalDate.of(2018, 10, 17))))
    }
}
